src/DynamoCore/Graph/Nodes/CustomNodes/Function.cs

```diff
@@ -444,6 +444,10 @@ public string InputSymbol
                             Parameter = new TypedParameter(name, type, defaultValue, null, comment);
                         }
                     }
+                    else
+                    {
+                        Parameter = new TypedParameter(name, type, defaultValue, null, comment);
+                    }
```

Comment:
This looks way more complicated than necessary. Seems like we can just do `type = identifierNode.datatype` right after the `name = identifierNode.Value` and then pull the `Parameter = new TypedParameter(name, type, defaultValue, null, comment);` out of all the cases to be after the `if (TryParse...`


                var type = TypeSystem.BuildPrimitiveTypeObject(PrimitiveType.Var);
                AssociativeNode defaultValue = null;

                string comment = null;

                if (!string.IsNullOrEmpty(name))
                {
                    // three cases:
                    //    x = default_value
                    //    x : type
                    //    x : type = default_value
                    IdentifierNode identifierNode;
                    if (TryParseInputExpression(inputSymbol, out identifierNode, out defaultValue, out comment))
                    {
                        name = identifierNode.Value;
                        type = identifierNode.datatype;

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            string warningMessage = String.Format(
                                Properties.Resources.WarningCannotFindType,
                                identifierNode.datatype.Name);
                            this.Warning(warningMessage);
                        }
                    }
                }

                Parameter = new TypedParameter(name, type, defaultValue, null, comment);

                OnNodeModified();

Reply:
@gregmarr we had other issues. earlier inputsymbol was used directly...but in the new serialization, it's not the case..
Below tasks will explain better...
https://jira.autodesk.com/browse/QNTM-3786
https://jira.autodesk.com/browse/QNTM-3872

Reply:
All I did was rearrange the current code slightly to eliminate duplicate code, and the unnecessary introduction of a local variable.

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            // warning
                            var ltype = identifierNode.datatype;
                            Parameter = new TypedParameter(name, ltype, defaultValue, null, comment);
                        }
                        else
                        {
                            type = identifierNode.datatype;
                            Parameter = new TypedParameter(name, type, defaultValue, null, comment);
                        }

The only difference between these two blocks (other than the warning) is that the first block creates the `ltype` local variable, and the second uses the type local variable.  Both create a TypedParameter with the exact same data.  The `type` local variable is never used again after this point.  Thus, there is no reason to create the local variable, and this code becomes

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            // warning
                        }

                        type = identifierNode.datatype;
                        Parameter = new TypedParameter(name, type, defaultValue, null, comment);

Then, that gives you 

                    if (TryParseInputExpression(inputSymbol, out identifierNode, out defaultValue, out comment))
                    {
                        name = identifierNode.Value;

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            // warning
                        }

                        type = identifierNode.datatype;
                        Parameter = new TypedParameter(name, type, defaultValue, null, comment);
                    }
                    else
                    {
                        Parameter = new TypedParameter(name, type, defaultValue, null, comment);
                    }

Now the last line of the `if` and `else` blocks are the same, so simplify that again

                    if (TryParseInputExpression(inputSymbol, out identifierNode, out defaultValue, out comment))
                    {
                        name = identifierNode.Value;

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            // warning
                        }

                        type = identifierNode.datatype;
                    }

                    Parameter = new TypedParameter(name, type, defaultValue, null, comment);

So now you have the outer if

                if (!string.IsNullOrEmpty(name))
                {
                    IdentifierNode identifierNode;
                    if (TryParseInputExpression(inputSymbol, out identifierNode, out defaultValue, out comment))
                    {
                        name = identifierNode.Value;

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            // warning
                        }

                        type = identifierNode.datatype;
                    }

                    Parameter = new TypedParameter(name, type, defaultValue, null, comment);
                }
                else
                {
                    Parameter = new TypedParameter(name, type, defaultValue, null, comment);
                }

Once again, the last lines of the two blocks are the same, so you can factor that out.

                if (!string.IsNullOrEmpty(name))
                {
                    IdentifierNode identifierNode;
                    if (TryParseInputExpression(inputSymbol, out identifierNode, out defaultValue, out comment))
                    {
                        name = identifierNode.Value;

                        if (identifierNode.datatype.UID == Constants.kInvalidIndex)
                        {
                            // warning
                        }

                        type = identifierNode.datatype;
                    }
                }

                Parameter = new TypedParameter(name, type, defaultValue, null, comment);

Then I just moved the assignment of `type` up under `name`.